From 0a06b9082c88ad7b5f992fdcec4851e2889b0f05 Mon Sep 17 00:00:00 2001
From: hcherndon <hcherndon@gmail.com>
Date: Wed, 2 Oct 2013 23:01:47 -0700
Subject: [PATCH] (Almura) Add Additional Methods

- Adds setMaxPlayers(int)
- Adds setMotd(String)
- Adds getPlayerList()

diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 65efa35..54b0280 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -1067,4 +1067,10 @@ public abstract class PlayerList {
     public void sendMessage(ChatMessage chatmessage) {
         this.sendMessage(chatmessage, true);
     }
+
+    // Almura Start
+    public void setMaxPlayers(int max){
+        this.maxPlayers = max;
+    }
+    // Almura End
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 7eea532..4ca487c 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1163,7 +1163,7 @@ public final class CraftServer implements Server {
     }
 
     public void setWhitelist(boolean value) {
-        playerList.hasWhitelist = value;
+        playerList.setHasWhitelist(value); // AEM
         console.getPropertyManager().a("white-list", value);
     }
 
@@ -1415,4 +1415,18 @@ public final class CraftServer implements Server {
         this.printSaveWarning = true;
         getLogger().log(Level.WARNING, "A manual (plugin-induced) save has been detected while server is configured to auto-save. This may affect performance.", warningState == WarningState.ON ? new Throwable() : null);
     }
+
+    // Almura Start
+    public DedicatedPlayerList getPlayerList(){
+        return this.playerList;
+    }
+
+    public void setMaxPlayers(int max) {
+        ((PlayerList)playerList).setMaxPlayers(max);
+    }
+
+    public void setMotd(String motd){
+        console.setMotd(motd);
+    }
+    // Almura End
 }
-- 
1.9.0.msysgit.0

