From 03de3db8dee537912461a71d3d62329b4ce3f236 Mon Sep 17 00:00:00 2001
From: Steven Downer <grinch@outlook.com>
Date: Fri, 4 Oct 2013 15:13:09 -0500
Subject: [PATCH] (Almura) Add ability to regenerate chunks without block
 updates


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 4076e6d..83776aa 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -193,7 +193,13 @@ public class CraftWorld implements World {
         return true;
     }
 
+    // Almura Start
     public boolean regenerateChunk(int x, int z) {
+        return regenerateChunk(x, z, true);
+    }
+
+    public boolean regenerateChunk(int x, int z, boolean refresh) {
+    // Almura End
         unloadChunk(x, z, false, false);
 
         world.chunkProviderServer.unloadQueue.remove(x, z);
@@ -208,7 +214,11 @@ public class CraftWorld implements World {
 
         chunkLoadPostProcess(chunk, x, z);
 
-        refreshChunk(x, z);
+        // Almura Start
+        if (refresh) {
+            refreshChunk(x, z);
+        }
+        // Almura End
 
         return chunk != null;
     }
-- 
1.9.0.msysgit.0

