From eb663085ec42ecafd464602bc14028cbd90cc692 Mon Sep 17 00:00:00 2001
From: Steven Downer <grinch@outlook.com>
Date: Thu, 17 Oct 2013 22:37:13 -0500
Subject: [PATCH] (Almura) Misc. Changes

- CraftOfflinePlayer: protected -> public

Signed-off-by: Steven Downer <grinch@outlook.com>

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
index dd6bcf9..d596e69 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftOfflinePlayer.java
@@ -26,7 +26,7 @@ public class CraftOfflinePlayer implements OfflinePlayer, ConfigurationSerializa
     private final CraftServer server;
     private final WorldNBTStorage storage;
 
-    protected CraftOfflinePlayer(CraftServer server, String name) {
+    public CraftOfflinePlayer(CraftServer server, String name) { // Almura - protected -> public
         this.server = server;
         this.name = name;
         this.storage = (WorldNBTStorage) (server.console.worlds.get(0).getDataManager());
-- 
1.9.0.msysgit.0

